<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python

In [3]:
connection = psycopg2.connect(
    dbname="project_sql",
    user="skillfactory",
    host="84.201.134.129",
    password="cCkxxLVrDE8EbvjueeMedPKt",
    port=5432
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [4]:
query_3_1 = f'''
SELECT COUNT(*) AS cnt
FROM public.vacancies;
'''

In [5]:
vacancies_count = pd.read_sql_query(query_3_1, connection)
vacancies_count

/tmp/ipython-input-3020254417.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_count = pd.read_sql_query(query_3_1, connection)


,cnt
0,49197


49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [6]:
query_3_2 = """
SELECT COUNT(*) AS cnt
FROM public.employers;
"""

employers_count = pd.read_sql_query(query_3_2, connection)
employers_count

/tmp/ipython-input-2856424049.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_count = pd.read_sql_query(query_3_2, connection)


,cnt
0,23501


23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [7]:
query_3_3 = """
SELECT COUNT(*) AS cnt
FROM public.areas;
"""

areas_count = pd.read_sql_query(query_3_3, connection)
areas_count

/tmp/ipython-input-263575252.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  areas_count = pd.read_sql_query(query_3_3, connection)


,cnt
0,1362


1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [8]:
query_3_4 = """
SELECT COUNT(*) AS cnt
FROM public.industries;
"""

industries_count = pd.read_sql_query(query_3_4, connection)
industries_count

/tmp/ipython-input-2509701710.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industries_count = pd.read_sql_query(query_3_4, connection)


,cnt
0,294


294

***

- В базе ~50 тыс. вакансий и ~23 тыс. работодателей — данных много, рынок довольно крупный.

- Есть 1362 региона и 294 сферы деятельности — структура довольно разнообразная и детализированная.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [10]:
query_4_1 = """
SELECT
    a.name AS area,
    COUNT(v.id) AS cnt
FROM public.vacancies v
JOIN public.areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC;
"""

vac_per_area = pd.read_sql_query(query_4_1, connection)
vac_per_area

/tmp/ipython-input-2006172836.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_per_area = pd.read_sql_query(query_4_1, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
764,Тарко-Сале,1
765,Новоаннинский,1
766,Бирск,1
767,Сасово,1


	area	cnt
0	Москва	5333


1	Санкт-Петербург	2851


2	Минск	2112


3	Новосибирск	2006


4	Алматы	1892


...	...	...


764	Тарко-Сале	1


765	Новоаннинский	1


766	Бирск	1


767	Сасово	1


768	Анна	1

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [11]:
query_4_2 = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE salary_from IS NOT NULL
   OR salary_to   IS NOT NULL;
"""

vac_with_salary = pd.read_sql_query(query_4_2, connection)
vac_with_salary

/tmp/ipython-input-3469404970.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_with_salary = pd.read_sql_query(query_4_2, connection)


,cnt
0,24073


24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [12]:
query_4_3 = """
SELECT
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to))   AS avg_salary_to
FROM public.vacancies;
"""

avg_salaries = pd.read_sql_query(query_4_3, connection)
avg_salaries

/tmp/ipython-input-1205970373.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salaries = pd.read_sql_query(query_4_3, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


71065.0

110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [13]:
query_4_4 = """
SELECT
    schedule,
    employment,
    COUNT(*) AS cnt
FROM public.vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC;
"""

vac_by_schedule_employment = pd.read_sql_query(query_4_4, connection)
vac_by_schedule_employment

/tmp/ipython-input-1267631970.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_by_schedule_employment = pd.read_sql_query(query_4_4, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


Полный день	Полная занятость	35367


1	Удаленная работа	Полная занятость	7802


2	Гибкий график	Полная занятость	1593


3	Удаленная работа	Частичная занятость	1312

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [14]:
query_4_5 = """
SELECT
    experience,
    COUNT(*) AS cnt
FROM public.vacancies
GROUP BY experience
ORDER BY cnt;
"""

vac_by_experience = pd.read_sql_query(query_4_5, connection)
vac_by_experience

/tmp/ipython-input-2350935260.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_by_experience = pd.read_sql_query(query_4_5, connection)


,experience,cnt
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


0	Более 6 лет	1337


1	Нет опыта	7197


2	От 3 до 6 лет	14511


3	От 1 года до 3 лет	26152

- Больше всего вакансий в крупных городах: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы.

- Примерно половина вакансий содержит хотя бы одну границу зарплатной вилки.

- Доминирует «полный день + полная занятость»; удалёнки тоже много, но заметно меньше.

- Основной спрос — на специалистов с опытом 1–3 и 3–6 лет, совсем без опыта вакансий значительно меньше.

***

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [16]:
query_5_1 = """
WITH emp_vac AS (
    SELECT
        e.id,
        e.name,
        COUNT(v.id) AS cnt
    FROM public.employers e
    LEFT JOIN public.vacancies v ON v.employer_id = e.id
    GROUP BY e.id, e.name
)
SELECT name, cnt, rn
FROM (
    SELECT
        name,
        cnt,
        ROW_NUMBER() OVER (ORDER BY cnt DESC) AS rn
    FROM emp_vac
) t
WHERE rn IN (1, 5);
"""

top_1_5_employers = pd.read_sql_query(query_5_1, connection)
top_1_5_employers

/tmp/ipython-input-371871357.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_1_5_employers = pd.read_sql_query(query_5_1, connection)


,name,cnt,rn
0,Яндекс,1933,1
1,Газпром нефть,331,5


0	Яндекс	1933


1	Газпром нефть	331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [52]:
query_5_2_extra = """
SELECT
    a.name AS area_name,
    COUNT(DISTINCT e.id) AS employers_cnt
FROM public.areas a
LEFT JOIN public.employers e ON e.area = a.id
LEFT JOIN public.vacancies v ON v.area_id = a.id
WHERE v.id IS NULL                 -- регионы, где нет вакансий
GROUP BY a.name
ORDER BY employers_cnt DESC
LIMIT 1;
"""

region_no_vac = pd.read_sql_query(query_5_2_extra, connection)
region_no_vac

/tmp/ipython-input-2031786972.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  region_no_vac = pd.read_sql_query(query_5_2_extra, connection)


,area_name,employers_cnt
0,Россия,410



Россия	410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [18]:
query_5_3 = """
SELECT
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS areas_cnt
FROM public.employers e
LEFT JOIN public.vacancies v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY areas_cnt DESC, employer_name;
"""

employer_regions = pd.read_sql_query(query_5_3, connection)
employer_regions.head()

/tmp/ipython-input-2178764199.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_regions = pd.read_sql_query(query_5_3, connection)


,employer_name,areas_cnt
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71


0	Яндекс	181


1	Ростелеком	152


2	Спецремонт	116


3	Поляков Денис Иванович	88


4	ООО ЕФИН	71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [20]:
query_5_4 = """
SELECT COUNT(*) AS employers_without_industry
FROM public.employers e
LEFT JOIN public.employers_industries ei
       ON ei.employer_id = e.id
WHERE ei.industry_id IS NULL;
"""

no_industry = pd.read_sql_query(query_5_4, connection)
no_industry

/tmp/ipython-input-2717353240.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  no_industry = pd.read_sql_query(query_5_4, connection)


,employers_without_industry
0,8419


8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [22]:
query_5_5 = """
WITH emp_ind AS (
    SELECT
        e.id,
        e.name,
        COUNT(ei.industry_id) AS industries_cnt
    FROM public.employers e
    JOIN public.employers_industries ei
         ON ei.employer_id = e.id
    GROUP BY e.id, e.name
    HAVING COUNT(ei.industry_id) = 4
)
SELECT name, industries_cnt
FROM emp_ind
ORDER BY name
OFFSET 2 LIMIT 1;   -- третья строка в алфавитном списке
"""

third_company_4_ind = pd.read_sql_query(query_5_5, connection)
third_company_4_ind

/tmp/ipython-input-1052440094.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  third_company_4_ind = pd.read_sql_query(query_5_5, connection)


,name,industries_cnt
0,2ГИС,4


2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [24]:
query_5_6 = """
SELECT COUNT(DISTINCT ei.employer_id) AS employers_cnt
FROM public.employers_industries ei
JOIN public.industries i
     ON i.id = ei.industry_id
WHERE i.name = 'Разработка программного обеспечения';
"""

soft_dev_employers = pd.read_sql_query(query_5_6, connection)
soft_dev_employers

/tmp/ipython-input-2088002063.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  soft_dev_employers = pd.read_sql_query(query_5_6, connection)


,employers_cnt
0,3553


3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [26]:
million_cities = [
    "Москва",
    "Санкт-Петербург",
    "Новосибирск",
    "Екатеринбург",
    "Казань",
    "Нижний Новгород",
    "Челябинск",
    "Самара",
    "Омск",
    "Ростов-на-Дону",
    "Уфа",
    "Красноярск",
    "Воронеж",
    "Пермь",
    "Волгоград",
    "Краснодар",
]

cities_list_sql = ", ".join([f"'{c}'" for c in million_cities])
cities_list_sql

"'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Самара', 'Омск', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Воронеж', 'Пермь', 'Волгоград', 'Краснодар'"

In [27]:
query_5_7 = f"""
WITH city_stats AS (
    SELECT
        a.name AS area_name,
        COUNT(v.id) AS vacancies_cnt
    FROM public.vacancies v
    JOIN public.employers e ON e.id = v.employer_id
    JOIN public.areas a     ON a.id = v.area_id
    WHERE e.name = 'Яндекс'
      AND a.name IN ({cities_list_sql})
    GROUP BY a.name
),
full_stats AS (
    SELECT
        area_name,
        vacancies_cnt,
        0 AS ord
    FROM city_stats

    UNION ALL

    SELECT
        'Total' AS area_name,
        SUM(vacancies_cnt) AS vacancies_cnt,
        1 AS ord
    FROM city_stats
)
SELECT
    area_name,
    vacancies_cnt
FROM full_stats
ORDER BY ord, area_name;
"""

yandex_million_cities = pd.read_sql_query(query_5_7, connection)
yandex_million_cities

/tmp/ipython-input-3881020624.py:35: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_million_cities = pd.read_sql_query(query_5_7, connection)


,area_name,vacancies_cnt
0,Волгоград,24.0
1,Воронеж,32.0
2,Екатеринбург,39.0
3,Казань,25.0
4,Краснодар,30.0
5,Красноярск,23.0
6,Москва,54.0
7,Нижний Новгород,36.0
8,Новосибирск,35.0
9,Омск,21.0


0	Волгоград	24.0


1	Воронеж	32.0


2	Екатеринбург	39.0


3	Казань	25.0


4	Краснодар	30.0


5	Красноярск	23.0


6	Москва	54.0


7	Нижний Новгород	36.0


8	Новосибирск	35.0


9	Омск	21.0


10	Пермь	25.0


11	Ростов-на-Дону	25.0


12	Самара	26.0


13	Санкт-Петербург	42.0


14	Уфа	26.0


15	Челябинск	22.0


16	Total	485.0

***

- Яндекс лидерует по числу вакансий


- Крупные работодатели размещают вакансии сразу в десятках регионов; Яндекс присутствует в очень большом числе регионов и в большинстве городов-миллионников.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [30]:
query_6_1 = """
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE LOWER(name) LIKE '%data%'
   OR LOWER(name) LIKE '%данн%';
"""

data_related_vac = pd.read_sql_query(query_6_1, connection)
data_related_vac

/tmp/ipython-input-3611135039.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_related_vac = pd.read_sql_query(query_6_1, connection)


,cnt
0,1771


1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [32]:
query_6_2 = """
SELECT COUNT(*) AS cnt
FROM public.vacancies v
WHERE (
        LOWER(v.name) LIKE '%data scientist%'
     OR LOWER(v.name) LIKE '%data science%'
     OR LOWER(v.name) LIKE '%исследователь данных%'
     OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
     OR LOWER(v.name) LIKE '%machine learning%'
     OR LOWER(v.name) LIKE '%машинн%обучен%'
      )
  AND (
        LOWER(v.name) LIKE '%junior%'
     OR v.experience = 'Нет опыта'
     OR v.employment = 'Стажировка'
      );
"""

junior_ds_vac = pd.read_sql_query(query_6_2, connection)
junior_ds_vac

/tmp/ipython-input-2994831506.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_ds_vac = pd.read_sql_query(query_6_2, connection)


,cnt
0,51


51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [34]:
query_6_3 = """
SELECT COUNT(*) AS cnt
FROM public.vacancies v
WHERE (
        LOWER(v.name) LIKE '%data scientist%'
     OR LOWER(v.name) LIKE '%data science%'
     OR LOWER(v.name) LIKE '%исследователь данных%'
     OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
     OR LOWER(v.name) LIKE '%machine learning%'
     OR LOWER(v.name) LIKE '%машинн%обучен%'
      )
  AND v.key_skills IS NOT NULL
  AND (
        v.key_skills ILIKE '%SQL%'
     OR v.key_skills ILIKE '%Postgres%'
      );
"""

ds_with_sql = pd.read_sql_query(query_6_3, connection)
ds_with_sql

/tmp/ipython-input-1190428252.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_with_sql = pd.read_sql_query(query_6_3, connection)


,cnt
0,229


229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [36]:
query_6_4 = """
SELECT COUNT(*) AS cnt
FROM public.vacancies v
WHERE (
        LOWER(v.name) LIKE '%data scientist%'
     OR LOWER(v.name) LIKE '%data science%'
     OR LOWER(v.name) LIKE '%исследователь данных%'
     OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
     OR LOWER(v.name) LIKE '%machine learning%'
     OR LOWER(v.name) LIKE '%машинн%обучен%'
      )
  AND v.key_skills ILIKE '%Python%';
"""

ds_with_python = pd.read_sql_query(query_6_4, connection)
ds_with_python

/tmp/ipython-input-1484372676.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_with_python = pd.read_sql_query(query_6_4, connection)


,cnt
0,357


357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [51]:
query_6_5 = """
SELECT
    ROUND(
        AVG(
            array_length(
                string_to_array(key_skills, E'\n'),
                1
            )
        )::numeric,
        2
    ) AS avg_skills_per_ds_vacancy
FROM public.vacancies v
WHERE (
        LOWER(v.name) LIKE '%data scientist%'
     OR LOWER(v.name) LIKE '%data science%'
     OR LOWER(v.name) LIKE '%исследователь данных%'
     OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
     OR LOWER(v.name) LIKE '%machine learning%'
     OR LOWER(v.name) LIKE '%машинн%обучен%'
      )
  AND key_skills IS NOT NULL;
"""

avg_skills_ds = pd.read_sql_query(query_6_5, connection)
avg_skills_ds

/tmp/ipython-input-2721417663.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills_ds = pd.read_sql_query(query_6_5, connection)


,avg_skills_per_ds_vacancy
0,1.0


1.0

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [49]:
query_6_6 = """
WITH ds_vacancies AS (
    SELECT
        v.experience,
        -- "средняя" зарплата по вакансии:
        -- если обе границы есть -> обычное среднее
        -- если одна граница NULL -> вторая подставится дважды, и среднее будет равно ей
        (
            COALESCE(salary_from, salary_to)
          + COALESCE(salary_to, salary_from)
        ) / 2.0 AS salary_mid
    FROM public.vacancies v
    WHERE (
            LOWER(v.name) LIKE '%data scientist%'
         OR LOWER(v.name) LIKE '%data science%'
         OR LOWER(v.name) LIKE '%исследователь данных%'
         OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
         OR LOWER(v.name) LIKE '%machine learning%'
         OR LOWER(v.name) LIKE '%машинн%обучен%'
        )
      AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
)
SELECT
    experience,
    ROUND(AVG(salary_mid)) AS avg_salary
FROM ds_vacancies
GROUP BY experience
ORDER BY experience;
"""

ds_salary_by_exp = pd.read_sql_query(query_6_6, connection)
ds_salary_by_exp

/tmp/ipython-input-3656247676.py:31: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_by_exp = pd.read_sql_query(query_6_6, connection)


,experience,avg_salary
0,Более 6 лет,157933.0
1,Нет опыта,74643.0
2,От 1 года до 3 лет,144233.0
3,От 3 до 6 лет,256454.0


In [50]:
query_6_6_1 = """
WITH ds_vacancies AS (
    SELECT
        v.experience,
        (
            COALESCE(salary_from, salary_to)
          + COALESCE(salary_to, salary_from)
        ) / 2.0 AS salary_mid
    FROM public.vacancies v
    WHERE (
            LOWER(v.name) LIKE '%data scientist%'
         OR LOWER(v.name) LIKE '%data science%'
         OR LOWER(v.name) LIKE '%исследователь данных%'
         OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
         OR LOWER(v.name) LIKE '%machine learning%'
         OR LOWER(v.name) LIKE '%машинн%обучен%'
        )
      AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
)
SELECT
    ROUND(AVG(salary_mid)) AS avg_salary_3_6
FROM ds_vacancies
WHERE experience = 'От 3 до 6 лет';
"""

ds_salary_3_6 = pd.read_sql_query(query_6_6_1, connection)
ds_salary_3_6

/tmp/ipython-input-2492472743.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_3_6 = pd.read_sql_query(query_6_6_1, connection)


,avg_salary_3_6
0,256454.0


0	Более 6 лет	157933.0


1	Нет опыта	74643.0


2	От 1 года до 3 лет	144233.0


3	От 3 до 6 лет	256454.0

***

Вакансий для Junior-DS совсем немного по сравнению с общим числом DS.

В большинстве DS-вакансий фигурируют Python, SQL/Postgres и ещё несколько обязательных скиллов — стек довольно стандартный.

Средняя зарплата заметно растёт с опытом: самые высокие уровни оплаты у дата-саентистов с опытом от 3 до 6 лет и выше.

## Общий вывод по проекту

По результатам детального анализа можно сделать несколько выводов:

* Наибольшее количество вакансий сосредоточено в крупных городах (Москва и другие крупные регионы находятся в топе по числу вакансий).
* Значительная доля вакансий содержит указание зарплатной вилки, но часть работодателей по-прежнему не раскрывает доход.
* Чаще всего встречается полный рабочий день и полная занятость, гибкие форматы и частичная занятость представлены заметно реже.
* По требуемому опыту рынок смещён в сторону вакансий «от 1 года до 3 лет» и «от 3 до 6 лет», предложений без опыта меньше.

В ходе проекта были изучены данные по вакансиям, работодателям и регионам.  
Удалось:

* оценить структуру базы (сколько вакансий, работодателей, регионов и сфер деятельности);
* понять, в каких регионах сосредоточено основное количество предложений;
* посмотреть на распределение вакансий по опыту, типу занятости и графику работы;
* разобраться, какие работодатели наиболее активны и как они распределяют свои вакансии по регионам;
* выделить сегмент вакансий, связанных с Data Science, и проанализировать требования и уровень оплаты труда для таких позиций.



In [45]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Топ-10 самых популярных навыков вообще

Анализ вакансий, связанных с Data Science, показывает, что:

* Количество вакансий, так или иначе связанных с данными, достаточно заметное, но настоящих DS-позиции (по заданному условию) существенно меньше общего числа IT-вакансий.
* Для начинающих (Junior) дата-саентистов вакансий меньше всего, и конкуренция за такие позиции, вероятно, выше.
* В большинстве DS-вакансий в качестве ключевых навыков фигурируют базы данных (SQL / Postgres) и язык Python — фактически это стандартный стек.
* В среднем в описании одной DS-вакансии указывается несколько ключевых навыков, что говорит о комплексных требованиях к кандидатам.
* Средняя зарплата растёт вместе с опытом: максимальные значения наблюдаются для диапазона «От 3 до 6 лет» и «Более 6 лет», что логично с точки зрения рынка труда.

In [55]:
query_7 = """
WITH skills AS (
    SELECT
        trim(s) AS skill
    FROM public.vacancies v,
         unnest(string_to_array(v.key_skills, E'\t')) AS s
    WHERE v.key_skills IS NOT NULL
)
SELECT
    skill,
    COUNT(*) AS cnt
FROM skills
GROUP BY skill
ORDER BY cnt DESC
LIMIT 10;
"""

ds_salary_3_6 = pd.read_sql_query(query_7, connection)
ds_salary_3_6

/tmp/ipython-input-3414160383.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_3_6 = pd.read_sql_query(query_7, connection)


,skill,cnt
0,SQL,6977
1,Git,4731
2,Английский язык,4525
3,Работа в команде,3923
4,Linux,3747
5,JavaScript,3625
6,Python,3445
7,Управление проектами,2764
8,1С программирование,2647
9,PostgreSQL,2452



SQL	6977


Git	4731


Английский язык	4525


Работа в команде	3923


Linux	3747


JavaScript	3625


Python	3445


Управление проектами	2764


1С программирование	2647


PostgreSQL	2452

Работодатели, у которых больше всего DS-вакансий, и их средняя DS-зарплата

In [57]:
query_8 = """
WITH ds_vacancies AS (
    SELECT
        v.id,
        v.employer_id,
        (
            COALESCE(v.salary_from, v.salary_to)
          + COALESCE(v.salary_to, v.salary_from)
        ) / 2.0 AS salary_mid
    FROM public.vacancies v
    WHERE (
            LOWER(v.name) LIKE '%data scientist%'
         OR LOWER(v.name) LIKE '%data science%'
         OR LOWER(v.name) LIKE '%исследователь данных%'
         OR (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%')
         OR LOWER(v.name) LIKE '%machine learning%'
         OR LOWER(v.name) LIKE '%машинн%обучен%'
        )
),
ds_emp_stats AS (
    SELECT
        e.name AS employer_name,
        COUNT(d.id) AS ds_cnt,
        ROUND(AVG(d.salary_mid)) AS avg_ds_salary
    FROM ds_vacancies d
    JOIN public.employers e ON e.id = d.employer_id
    GROUP BY e.name
)
SELECT
    employer_name,
    ds_cnt,
    avg_ds_salary
FROM ds_emp_stats
WHERE ds_cnt >= 3         -- можно порог поменять
ORDER BY ds_cnt DESC, avg_ds_salary DESC
LIMIT 20;
"""

ds_salary_3_6 = pd.read_sql_query(query_8, connection)
ds_salary_3_6

/tmp/ipython-input-868063873.py:39: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary_3_6 = pd.read_sql_query(query_8, connection)


,employer_name,ds_cnt,avg_ds_salary
0,СБЕР,39,NaN
1,Bell Integrator,25,NaN
2,Банк ВТБ (ПАО),18,NaN
3,VK,15,NaN
4,Positive Technologies,11,NaN
5,EvenBet Gaming,9,NaN
6,Яндекс,9,NaN
7,МегаФон,8,NaN
8,Andersen,7,268863.0
9,Ozon,6,NaN


0	СБЕР	39	NaN

1	Bell Integrator	25	NaN

2	Банк ВТБ (ПАО)	18	NaN

3	VK	15	NaN

4	Positive Technologies	11	NaN

5	EvenBet Gaming	9	NaN

6	Яндекс	9	NaN

7	МегаФон	8	NaN

8	Andersen	7	268863.0

9	Ozon	6	NaN

10	МТС	6	NaN

11	Бэнкс Софт Системс	6	NaN

12	Иннотех, Группа компаний	6	NaN

13	Контур	6	NaN

14	Газпромбанк	5	NaN

15	2ГИС	5	NaN

16	Лента, федеральная розничная сеть	5	NaN

17	Лига Цифровой Экономики	5	NaN

18	Спортмастер	5	NaN

19	inDriver	5	NaN


In [44]:
#connection.close()